sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamTableFunctionScanRel.java

```diff
@@ -166,6 +247,54 @@ public void processElement(ProcessContext c) {
     }
   }
 
+  private static class SlidingWindowDoFn extends DoFn<Row, Row> {
+    private final int windowFieldIndex;
+    private final SlidingWindows windowFn;
+    private final Schema outputSchema;
+
+    public SlidingWindowDoFn(SlidingWindows windowFn, int windowFieldIndex, Schema schema) {
+      this.windowFn = windowFn;
+      this.windowFieldIndex = windowFieldIndex;
+      this.outputSchema = schema;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      Row row = c.element();
+      Collection<IntervalWindow> windows =
+          windowFn.assignWindows(row.getDateTime(windowFieldIndex).toInstant());
+      for (IntervalWindow window : windows) {
+        Row.Builder builder = Row.withSchema(outputSchema);
+        builder.addValues(row.getValues());
+        builder.addValue(window.start());
+        builder.addValue(window.end());
+        c.output(builder.build());
+      }
+    }
+  }
+
+  private static class SessionWindowDoFn extends DoFn<KV<String, Iterable<Row>>, Row> {
+    private final Schema outputSchema;
+
+    public SessionWindowDoFn(Schema schema) {
+      this.outputSchema = schema;
+    }
+
+    @ProcessElement
+    public void processElement(
+        @Element KV<String, Iterable<Row>> element, BoundedWindow window, OutputReceiver<Row> out) {
+      IntervalWindow intervalWindow = (IntervalWindow) window;
+      for (Row cur : element.getValue()) {
```

Comment:
This is going to iterate over every element in the window, that won't work. You might be able to make something that works reasonably well with `CombineFn`. (I don't think this is something we will figure out in PR comments.)

Reply:
See scalability comment above. Because GBK works on per-key and per-window basis, and I used "dummy" key above, thus this ParDo will go through every element in a session window. If that window is large, this could cause a problem. (Hot key is always a concern for any GBK).

Logged https://jira.apache.org/jira/browse/CALCITE-4051 for future improvement.